### PR TITLE
[IMPROVED] KeyValue documentation

### PR DESCRIPTION
**Pull request by piotrpio**
Signed-off-by: Piotr Piotrowski <piotr@synadia.com>

### CODE REVIEW

File: jetstream/kv.go [hunk 1]
```diff
@@ -27,149 +27,287 @@ import (
...
+	// - Purge all delete markers
+	// - Close the KeyValue store
 	KeyValue interface {
-		// Get returns the latest value for the key.
+		// Get returns the latest value for the key. If the key does not exist,
+		// ErrKeyNotFound will be returned.
 		Get(ctx context.Context, key string) (KeyValueEntry, error)
-		// GetRevision returns a specific revision value for the key.
+
+		// GetRevision returns a specific revision value for the key. If the key
```
Comment (ripienaar) [inline jetstream/kv.go:76]: also if the revision isnt found?

**Review by ripienaar — APPROVED** [review]
LGTM except for one question

File: jetstream/README.md [hunk 2]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+### Basic usage of KV bucket
+
+The most basic usage of KV bucket is to create or retrieve a bucket and perform
+basic CRUD operations on keys.
+
+```go
+js, _ := jetstream.New(nc)
+ctx := context.Background()
+
+// Create a new bucket. Bucket name is required and has to be unique within a stream.
```
Comment (Jarema) [inline jetstream/README.md:638]: The bucket is a stream, so not sure what it means?

File: jetstream/README.md [hunk 3]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+
+// Get an entry for a given key
+// Entry contains key/value, but also metadata (revision, timestamp, etc.)) 
+entry, _ := kv.Get(ctx, "sue.color")
+
+// Prints `sue.color @ 1 -> "blue"`
+fmt.Printf("%s @ %d -> %q\n", entry.Key(), entry.Revision(), string(entry.Value()))
+
+// Update a value for a given key
+// Update will fail if the key does not exist or the revision is not up to date
```
Comment (Jarema) [inline jetstream/README.md:653]: hm, "or the revision changed"?

File: jetstream/README.md [hunk 4]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+kv.Update(ctx, "sue.color", []byte("red"), 1)
+
+// Create will fail if the key already exists
+_, err := kv.Create(ctx, "sue.color", []byte("purple"))
+fmt.Println(err) // prints `nats: key exists`
+
+// Delete a value for a given key
```
Comment (Jarema) [inline jetstream/README.md:660]: Mention its non-destructive

File: jetstream/README.md [hunk 5]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+changes occur.
+
+Watcher supports several configuration options:
+
+- `IncludeHistory` will have the key watcher send all historical values
+for each key (up to KeyValueMaxHistory).
+- `IgnoreDeletes` will have the key watcher not pass any keys with
+delete markers.
+- `UpdatesOnly` will have the key watcher only pass updates on values
+(without latest values when started).
```
Comment (Jarema) [inline jetstream/README.md:687]:
```suggestion
(without values already present when starting).
```

File: jetstream/README.md [hunk 6]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+- `MetaOnly` will have the key watcher retrieve only the entry meta
+data, not the entry value.
```
Comment (Jarema) [inline jetstream/README.md:689]: metadata as one word?

File: jetstream/README.md [hunk 7]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+```go
+js, _ := jetstream.New(nc)
+ctx := context.Background()
+kv, _ := js.CreateKeyValue(ctx, jetstream.KeyValueConfig{Bucket: "profiles"})
+
+kv.Put(ctx, "sue.color", []byte("blue"))
+
+// A watcher can be created to watch for changes on a given key or the whole bucket
+// Watcher will receive a notification on a channel when a change occurs
+// By default, watcher will return initial values for all matching keys.
```
Comment (Jarema) [inline jetstream/README.md:702]:
```suggestion
// By default, watcher will return most recent values for all matching keys.
```

File: jetstream/README.md [hunk 8]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+// Watcher can be configured to only return updates by using jetstream.UpdatesOnly() option.
+watcher, _ := kv.Watch(ctx, "sue.*")
+defer watcher.Stop()
+
+kv.Put(ctx, "sue.age", []byte("43"))
+kv.Put(ctx, "sue.color", []byte("red"))
+
+// First, the watcher sends initial values for all matching keys
```
Comment (Jarema) [inline jetstream/README.md:710]:
```suggestion
// First, the watcher sends last known values for all matching keys
```

Comment (Jarema) [reply jetstream/README.md:710]: little different naming idea than the previous one

File: jetstream/README.md [hunk 9]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+// In this case, it will send a single entry for `sue.color`.
+entry := <-watcher.Updates()
+// Prints `sue.color @ 1 -> "blue"`
+fmt.Printf("%s @ %d -> %q\n", entry.Key(), entry.Revision(), string(entry.Value()))
+
+// After all initial values have been sent, watcher will send nil on the channel.
```
Comment (Jarema) [inline jetstream/README.md:716]:
```suggestion
// After all current values have been sent, watcher will send nil on the channel.
```

File: jetstream/README.md [hunk 10]
```diff
@@ -603,6 +609,193 @@ ackF, err = js.PublishAsync("ORDERS.new", []byte("hello"))
...
+ctx := context.Background()
+kv, _ := js.CreateKeyValue(ctx, jetstream.KeyValueConfig{Bucket: "profiles"})
+
+kv.Put(ctx, "sue.color", []byte("blue"))
+kv.Put(ctx, "sue.age", []byte("43"))
+kv.Put(ctx, "bucket", []byte("profiles"))
+
+// Purge will remove all keys from a bucket
+// The latest revision of each key will be kept
+// with a delete marker, all previous revisions will be removed
```
Comment (Jarema) [inline jetstream/README.md:771]: I would make a distinction between purge and delete clear in a sense that one is destructive, other is not.

**Review by Jarema — CHANGES_REQUESTED** [review]
Great addition!

Some comments and suggestions. I know you explain those mentioned ones clearly in the methods/interfaces names, but still, might be good to improve.

File: jetstream/README.md (hunk 2 above)
Comment (piotrpio) [reply jetstream/README.md:638]: Yeah, that does not make sense - it was meant to be "within a JetStream account".

File: jetstream/README.md (hunk 8 above)
Comment (piotrpio) [reply jetstream/README.md:710]: I think I prefer "most recent", "last known" sounds a little vague to me.

File: jetstream/README.md (hunk 10 above)
Comment (piotrpio) [reply jetstream/README.md:771]: I will add further clarification in method descriptions, only slight refinement here. I wouldn't like to add too much commentary in code examples.

**Review by Jarema — APPROVED** [review]
LGTM!

**Pull request merged**